### PR TITLE
109 Fixes

### PR DESCRIPTION
**Pull request by JohnHadish**
Solved issue #109 

## Description

This pull request solves the confusing naming structure of the prefix. The issue was that GEMmaker originally only did Hisat2 alignment. For this, it made since to use the prefix. For Salmon and Kallisto, however, it does not make since to use a prefix since both of these tools only use one argument. This is addressing an issue a user had where they became frustrated due to not understanding the prefix option. They  were only using Kallisto.

New Format:
```
    // Reference paths (note, if you are only using one alignment tool, i.e. kallisto, you do not need references for the other alignment tools)

    // Point to the salmon index directory
    salmon_reference_file_fullpath = "${baseDir}/examples/reference/CORG.transcripts.Salmon.indexed"

    // Point to the kallisto index file
    kallisto_reference_file_fullpath = "${baseDir}/examples/reference/CORG.transcripts.Kallisto.indexed"

    // To run Hisat2, must have the ".gtf" file and ".ht2" build files.
    hisat_reference_path = "${baseDir}/examples/reference/"
    hisat_reference_prefix = "CORG"
```

## Additional Minor changes in this pull request:

Removed all:
```
_vs_${params.input.reference_prefix}
```
as they are unnecessary and make the code less readable.

Added code that will state the alignment tool being used at the top of the nextflow run, and throw an error if an incorrect argument is given to the ```params.software.alignment```
example:
```
Software Parameters:
--------------------
  Trimmomatic clip path:      /home/john/Documents/GEMmaker/files/fasta_adapter.txt
  Trimmomatic minimum ratio:  0.7
  Alignment tool: Salmon
```

## Testing?

Travis CI should run properly, can also test locally to make sure that Salmon and Kallisto are running properly.

### CODE REVIEW

File: bin/create-gem.py [hunk 1]
```diff
@@ -54,7 +54,7 @@
 
   # Get the sample name from the file name.
   file_basename = os.path.basename(result)
-  sample_name = file_basename.split('_vs_')[0]
+  sample_name = file_basename.split('.')[0]
```
Comment (spficklin) [inline bin/create-gem.py]: Hey John, why this change?

**Comment by spficklin** [comment]
@JohnHadish and @bentsherman I think we need similar fixes in the main.nf to also occur in the main-e2e.nf

**Comment by bentsherman** [comment]
It seems like it would make more sense to have separate subdirectories for hisat2 / salmon / kallisto. After all it looks like there are multiple input files associated with each tool. I think that would make the input settings much cleaner as well:
```
reference_path_salmon = "${baseDir}/examples/reference/CORG.transcripts.Salmon.indexed"
reference_path_kallisto = "${baseDir}/examples/reference/CORG.transcripts.Kallisto.indexed"
reference_path_hisat = "${baseDir}/examples/reference/"

reference_prefix = "CORG"
```

My issue with it right now is that this reference directory seems very unorganized, and the way that nextflow consumes these inputs is very different for each aligner. If we could use a more uniform approach for providing reference files, I think that would make the user's job much easier.

**Comment by spficklin** [comment]
I'm all for improved organization! Could we put hisat indexes in their own directory as well for consistency?

File: bin/create-gem.py (hunk 1 above)
Comment (JohnHadish) [reply bin/create-gem.py]: I didn't want the user to have to input a reference name if they were using Kallisto or Salmon. I thought that it would be confusing. I also thought that having this additional "_vs_" doesn't really make since to have. The user knows what they are aligning to, and including this information in the title seemed redundant.

**Comment by JohnHadish** [comment]
There is only 1 input file for Salmon and Kallisto. This is why I thought it would be easier to just use the path to the reference. Do we think it would be easier to just make directories to put one reference file in? Either way is fine by me.

In reference to 

> @JohnHadish and @bentsherman I think we need similar fixes in the `main.nf` to also occur in the `main-e2e.nf`



> @JohnHadish and @bentsherman I think we need similar fixes in the `main.nf` to also occur in the `main-e2e.nf`

I have implemented this as well

**Comment by spficklin** [comment]
> I have already implemented this fix in the main-e2e.nf as well

Oh, great! sorry I missed that.  

What if we do the following:
1. Create separate sections for each alignment tool.
2. Get rid of the `reference_path` variable.  It's a pain to make the GTF, hisat2 indexes and genome reference all have the same prefix.  I have examples of genomes where the assembly and the annotations are different versions and have different prefexes, so to rename the GTF file to match the assembly name is incorrect.  
3.  Get rid of the `params.software.alignment` variable.  I think it's more intuitive if we have an "enable" flag for each of the tools.  And I think it would allow GEMmaker to run all three tools simultaneously as well.  Although, we should probably put the name of the tool in the name of the output files so we don't have conflicts.
4.  Get rid of the `params.input.reference_prefix`.  It's really not needed, and we can use the `params.project.machine_name` in place of the reference prefix for all output files.

For example:

```
params {
  input {
    hisat2 {
      enable = "yes"
      indexes_dir = "${baseDir}/examples/reference/"
    }
    salmon {
      enable = "no"
      index_file = "${baseDir}/examples/reference/CORG.transcripts.Salmon.indexed"
    }  
    kallisto {
      enable = "no"
      index_file = "${baseDir}/examples/reference/CORG.transcripts.Kallisto.indexed"
    }
  }
}
```

**Comment by bentsherman** [comment]
@spficklin I like your ideas a lot. This should also solve #100.

The config file supports booleans, so you can do something like:
```
hisat2 {
  enabled = true
}
```

File: bin/create-gem.py (hunk 1 above)
Comment (spficklin) [reply bin/create-gem.py]: I think we need to keep the _vs_{reference}.  If someone runs the same samples against different reference versions and decides to keep and share the BAM and other output files then it can get confusing which reference the reads were aligned to.

Comment (spficklin) [reply bin/create-gem.py]: Alternatively, if we follow my suggestion to get rid of the reference_prefix then we would use the machine_name for the project rather than the `_vs_{reference}`.  That I think would do the same thing but if we do that we wouldn't need to use the `_vs_` string.

**Comment by spficklin** [comment]
Okay, I went through and fixed the config file so that the index settings are per tool and there is an `enable` setting for each tool.  I was hoping we could turn all the tools on and GEMmaker could run all of them at once.  But that can't happen right now as we'd need better logic for when its time to move the next sample.  So, for now, you can only enable one at a time.  

I ran out of time to update the documentation and the main-e2e.nf so if someone wouldn't mind fixing those that would be great. 

Also, @JohnHadish I think we need to put the `_vs_` back in for the reasons I mention above.

**Comment by spficklin** [comment]
Oh, but if we put the `_vs_` back in we'll have to use something else to identify the assembly name and version.  Maybe we should add some new parameters such as 
`genome_assembly_name` and use that instead.  We'd need some way to distinguish what the assembly the sample was aligned to.

**Comment by spficklin** [comment]
@JohnHadish and @bentsherman . I have finished up all my edits.  Specifically, I merged in the changes from the last PR #119, I migrated all of the config setting updates to the `main-e2e.nf`, I updated the documentation to match the changes and I put back the `_vs_` in some of the intermediate files.  

Can both of you test it to make sure all of my late night changes haven't flubbed anything up?

**Comment by JohnHadish** [comment]
@spficklin 
It runs fine, but I am confused why we have two very similar parameters in the nextflow.config:

<pre>

    // The unique name for the genome reference assembly. It must not contain
    // spaces or special characters, only alphanumeric characters and underscores.
    <b>reference_name = "CORG"</b>

    // Gene quantification tool inputs. Note, if you are only using one alignment
    // tool, i.e. kallisto, you do not need references for the other alignment tools
    hisat2 {
      enable = true
      index_dir = "${baseDir}/examples/reference/CORG.transcripts.Hisat2.indexed/"
      <b>index_prefix = "CORG"</b>
      gtf_file = "${baseDir}/examples/reference/CORG.gtf"
    }
</pre>

The first of these seems to defeat the initial purpose of this pull request, as the prefix is not needed for Salmon and Kallisto, and tends to just cause confusion. 

I like the new format of running the different alignment tools besides this issue.
